clip_retrieval/clip_inference/slurm_distributor.py

```diff
@@ -0,0 +1,195 @@
+"""Distribute work using SLURM"""
+
+import os
+import time
+import subprocess
+from multiprocessing.pool import ThreadPool
+
+
+class SlurmDistributor:
+    """distribute work across a collection of slurm nodes"""
+
+    def __init__(self, tasks, worker_args, slurm_args):
+        self.worker_args = worker_args
+        self.slurm_args = slurm_args
+        self.nodes = self.slurm_args.pop("nodes")
+
+        self.job_timeout = slurm_args.pop("job_timeout")
+
+        # calculate world info for distributing work, assume 1 GPU/node
+        self.tasks = tasks
+        self.num_tasks = len(self.tasks)
+        self.tasks_per_node = self.num_tasks // self.nodes
+
+        if self.tasks_per_node <= 0:
+            print("There are more nodes than tasks...reducing the number of requested nodes.")
+
+            while self.tasks_per_node <= 0:
+                # reduce the number of nodes by one until we no longer have an excess
+                self.nodes -= 1
+                self.tasks_per_node = self.num_tasks // self.nodes
+
+            new_worker_count = self.nodes
+            print(f"Now using only: {new_worker_count} workers")
+
+    def __call__(self):
+        """
+        Distribute task amongst multiple workers that will be started with sbatch.
+        """
+        cache_path = self.slurm_args.pop("cache_path")
+        if cache_path is None:
+            cache_path = os.path.expanduser("~/.cache")
+        if not os.path.exists(cache_path):
+            os.makedirs(cache_path, exist_ok=True)
+
+        task_assignments = {node_id: self._get_worker_tasks(node_id) for node_id in range(self.nodes)}
+
+        # create the sbatch files to run
+        sbatch_files = []
+        for node_id, tasks in task_assignments.items():
+            # save the sbatch file to the cache
+            sbatch_script_path = os.path.join(cache_path, f"sbatch_{node_id}.script")
+            sbatch_output_path = os.path.join(cache_path, f"sbatch_{node_id}.out")
+
+            with open(sbatch_script_path, "w", encoding="utf-8") as sbatch_file:
+                sbatch_file.write(
+                    self._generate_sbatch(
+                        tasks=tasks, output_file=sbatch_output_path, cache_folder=cache_path, **self.slurm_args
+                    )
+                )
+
+            sbatch_files.append(sbatch_script_path)
+
+        # create a thread group to manage all the jobs we are about to start
+        all_results = {}
+        with ThreadPool(self.nodes) as p:
```

Comment:
Could handle failure here

Could also be possible to start a single pool for N nodes

Also need to rename nodes by jobs